cmake/Gems.cmake

```diff
@@ -252,3 +252,96 @@ function(ly_enable_gems_delayed)
+        set(ly_is_gem_available_VARIANT "Clients")
+    endif()
+
+    if (NOT ly_is_gem_available_VALUE_IF_FOUND)
+        set(ly_is_gem_available_VALUE_IF_FOUND TRUE)
+    endif()
+
+    if (NOT ly_is_gem_available_VALUE_IF_NOT_FOUND)
+        set(ly_is_gem_available_VALUE_IF_NOT_FOUND FALSE)
+    endif()
+
+    if (NOT ly_is_gem_available_RETURN_VARIABLE)
+        message(WARNING "Call to 'ly_is_gem_available' without specifying the RETURN_VARIABLE argument. The call was ignored.")
+        return()
+    endif()
+
+    get_property(ly_delayed_enable_gems GLOBAL PROPERTY LY_DELAYED_ENABLE_GEMS)
+    foreach(project_target_variant ${ly_delayed_enable_gems})
+        # we expect a colon separated list of
+        # PROJECT_NAME,target_name,variant_name
+        string(REPLACE "," ";" project_target_variant_list "${project_target_variant}")
+        list(LENGTH project_target_variant_list project_target_variant_length)
+        if(project_target_variant_length EQUAL 0)
+            continue()
+        endif()
+        
+        if(NOT project_target_variant_length EQUAL 3)
+            message(FATAL_ERROR "Invalid specification of gems, expected 'project','target','variant' and got ${project_target_variant}")
+        endif()
+
+        list(POP_BACK project_target_variant_list variant)
+        list(POP_BACK project_target_variant_list target)
+        list(POP_BACK project_target_variant_list project)
+
+        # If the current project variant is not the one we want.
+        if (NOT variant STREQUAL ${ly_is_gem_available_VARIANT})
+            continue()
+        endif()
+
+        get_property(gem_dependencies GLOBAL PROPERTY LY_DELAYED_ENABLE_GEMS_"${project_target_variant}")
+        if (NOT gem_dependencies)
+            # Continue to the next iteration loop regardless as there are no gem dependencies
+            continue()
+        endif()
+
+        if (NOT TARGET ${target})
+            message(FATAL_ERROR "ly_enable_gems specified TARGET '${target}' but no such target was found.")
+        endif()
+
+        # Fetch for the searched gem dependency.
+        foreach(gem_name ${gem_dependencies})
+            # the gem name may already have a namespace.  If it does, we use that one
+            ly_strip_target_namespace(TARGET ${gem_name} OUTPUT_VARIABLE unaliased_gem_name)
+            if (${unaliased_gem_name} STREQUAL ${gem_name})
+                # if stripping a namespace had no effect, it had no namespace
+                # and we supply the default Gem:: namespace.
+                set(gem_name_with_namespace Gem::${gem_name})
+            else()
+                # if stripping the namespace had an effect then we use the original
+                # with the namespace, instead of assuming Gem::
+                set(gem_name_with_namespace ${gem_name})
+            endif()
+            
+            # if the target exists, add it.
+            if (TARGET ${gem_name_with_namespace}.${variant})
```

Comment:
With your explanation I've done some further tests, and yes you are right in some case we have false positives, but I don't really able to know what is the real cause.

I assume that I may not fully understand the way O3DE is fetching gems 😅, but this is a list of tests I've made using this `enable_gems.cmake` file:

```cmake
set(ENABLED_GEMS
    SS_EXPERIMENTS
    Atom
    AudioSystem
    AWSCore
    CameraFramework
    DebugDraw
    EditorPythonBindings
    EMotionFX
    GameState
    ImGui
    LandscapeCanvas
    LyShine
    Multiplayer
    PhysX
    PrimitiveAssets
    SaveData
    ScriptCanvasPhysics
    ScriptEvents
    StartingPointInput
    SSBehaviorTree
    WhiteBox
    TextureAtlas
    Atom_AtomBridge
)
```

1. Called that function in the EMotionFX gem to see if the WhiteBox gem is available
```cmake
ly_is_gem_available(
    NAME WhiteBox
    RETURN_VARIABLE WHITEBOX_AVAILABLE
)
message( STATUS "EMOTIONFX - Is WhiteBox Available?: ${WHITEBOX_AVAILABLE}")
```
As we can expect and following your explanation EMotionFX is saying that WhiteBox is not available.

2. Moved up the `Gems/WhiteBox` path in the `engine.json` file to the top of the array, so it's coming before `Gems/EmotionFX`. Then I've cleaned the CMake cache and reconfigured the project, EMotionFX is always saying that WhiteBox is not available.

3. Then I've tried to move the declaration of WhiteBox to be before EMotionFX in the `enabled_gems.cmake` file, even like this EMotionFX is always saying that WhiteBox is not available 🤔

So I've finally wanted to see what are the available gems at the time of the registration of EMotionFX, with this code:

```cmake
if (TARGET ${gem_name_with_namespace}.${variant})
    message(STATUS "${gem_name} is available and registered.")
    if (gem_name STREQUAL ${ly_is_gem_available_NAME})
        set("${ly_is_gem_available_RETURN_VARIABLE}" "${ly_is_gem_available_VALUE_IF_FOUND}" PARENT_SCOPE)
        return()
    endif()
else()
    message(STATUS "${gem_name} is available but not yet registered.")
endif()
```

And this is the output:

```txt
[cmake] -- SS_EXPERIMENTS is available and registered.
[cmake] -- WhiteBox is available but not yet registered.
[cmake] -- SSBehaviorTree is available but not yet registered.
[cmake] -- Atom is available and registered.
[cmake] -- AudioSystem is available and registered.
[cmake] -- AWSCore is available and registered.
[cmake] -- CameraFramework is available and registered.
[cmake] -- DebugDraw is available and registered.
[cmake] -- EditorPythonBindings is available but not yet registered.
[cmake] -- EMotionFX is available but not yet registered.
[cmake] -- GameState is available but not yet registered.
[cmake] -- ImGui is available but not yet registered.
[cmake] -- LandscapeCanvas is available but not yet registered.
[cmake] -- LyShine is available but not yet registered.
[cmake] -- Multiplayer is available but not yet registered.
[cmake] -- PhysX is available but not yet registered.
[cmake] -- PrimitiveAssets is available but not yet registered.
[cmake] -- SaveData is available but not yet registered.
[cmake] -- ScriptCanvasPhysics is available but not yet registered.
[cmake] -- ScriptEvents is available but not yet registered.
[cmake] -- StartingPointInput is available but not yet registered.
[cmake] -- TextureAtlas is available but not yet registered.
[cmake] -- Atom_AtomBridge is available and registered.
[cmake] -- EMOTIONFX - Is WhiteBox Available?: FALSE
```

As we can see, all gems used for the project are _available_ at the time EMotionFX do the check, but some of them are just not yet _registered_.

So for that I have a question @amznestebanpapp, Is it necessary for Gem2 to be registered at the time Gem1 is asking for his availability? If Gem1 know that Gem2 is available, is it not meaning that Gem2 will be surely registered sometime?

Reply:
For now, removing the check `if (TARGET ${gem_name_with_namespace}.${variant})` have fixed the issue...

Reply:
You can follow the walking logic in the root CMakeLIsts.txt lines 82-98, that "external directory walking" is where is walking through the gems. That list is read in `add_engine_json_external_subdirectories` and should follow the order in the json file.

Taking a step back, I think this is overcomplicating things.
If a Gem wants to use another Gem, it needs to have it as a runtime dependency (RUNTIME_DEPENDENCIES). 
Are you trying to make it "optional", e.g. you want to make it a RUNTIME_DEPENDENCY only if the other gem is in the build tree?

Reply:
> Are you trying to make it "optional", e.g. you want to make it a RUNTIME_DEPENDENCY only if the other gem is in the build tree?

Yes, this is one of the goals. I'm currently building a BehaviorTree engine which will implement some O3DE features plus some gems features into behavior tree nodes.

For example, the case of the EMotionFX gem where I want to enable some nodes only when the project depends on that gem.

I've ended up to write this function because I haven't wanted to create many gems for the same purpose and just because of the dependencies (ex: BehaviorTree.Core gem, BehaviorTree.EMotionFX gem, BehaviorTree.WWiseAudio gem, etc...), so if I've chosen to find a way to configure the gem itself using CMake directly, so whatever gem is added/removed, a CMake configure happen and my gem is updated with the new dependencies.

For now, by removing the check on the target make the thing to work, at least for my case.

Reply:
So it sounds like, you are looking more to know if a particular gem has been loaded.
Based off of that, you are looking to add additional nodes to the Behavior Tree Model.

This could be done by using the ModuleManagerRequestBus EnumerateModules EBus event to determine which gems are loaded. Here is the link to [ModuleManagerBus.h](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Module/ModuleManagerBus.h#L75-L88) header.
The ModuleManagerBus is implemented in the [ModuleManager](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Module/ModuleManager.h#L136-L141) class

An example of enumerating over the list of loaded gems and determining the gem module target name can be found in the SerializeContextTools Converter.cpp [ConvertModuleComponents](https://github.com/o3de/o3de/blob/stabilization/2106/Code/Tools/SerializeContextTools/Converter.cpp#L616-L688) function

Reply:
Thanks for the links @lumberyard-employee-dm I will check that 👍🏾. But a quick question, to do that I don't need to add these gems in the RUNTIME_DEPENDENCIES/BUILD_DEPENDENCIES of my gem?

Because I not only have to know if a gem is loaded, but I also need to consume it (by including some of their header files). This is mainly why I wanted to switch on/off a gem-based node at the project generation.

Reply:
Well RUNTIME_DEPENDENCIES will never propagate information like include directories, compile definitions and link targets to the target it is tied to.

What it sounds like is the functionality you need is as follows:
After all the CMake targets have been declared and defined by calling `add_subdirectory` on each of their CMakeLists.txt and invoking the appropriate `add_library/add_executable` calls through `ly_add_target`. you need to know a specified set of gems targets that the project depends on.

Based off of that information, the BehaviorTree gem needs a post-cmake configure hook where it adds the dependencies of specified "gem set" that are enabled to the BehaviorTree gem target, so that it is' include directories are available.

If you went with that approach though, that would require `#ifdef` littered through out the BehaviorTree gem to selectively include files from specific gem targets and invoke the functions you need on them.

I think an alternative approach is to data-drive the registration of Nodes from other gems via the Settings Registry and avoid the need to have a build dependency on other gems directly.

The Settings Registry is a data store of settings that is globally access across the entire engine.
It can be populated by .setreg files which are [JSON Merge Patch](https://datatracker.ietf.org/doc/html/rfc7386) formatted files or .setregpatch files which are [JSON Patch](https://datatracker.ietf.org/doc/html/rfc6902/) files.
Using each of those algorithms the values within those files are merged into the Settings Registry.
Examples of .setreg files can be found within the [<engine-root>/Registry](https://github.com/o3de/o3de/tree/development/Registry) directory.

A gem can contain a Registry folder right underneath the <gem-root>, that will get loaded and merged into the Settings Registry before any gems are loaded.

So one way to implement this is to have the BehaviorTree gem contain a `Registry`  directory containing .setreg file describing the list of nodes to populate from each gem
```
BehaviorTree/
   Code/...
   Registry/
      behavior_tree_nodes.emotionfx.setreg
      behavior_tree_nodes.audioenginewwise.setreg
      behavior_tree_nodes.physx.setreg
```

The Settings Registry supports filtering specific .setreg/.setregpatch files within a folder using it's specialization tag system.
We don't have the Settings Registry documentation on the o3de.org site yet, so the doc comments on the interface class will have to suffice for now. Here is the interface of [MergeSettingsFolder](https://github.com/o3de/o3de/blob/stabilization/2106/Code/Framework/AzCore/AzCore/Settings/SettingsRegistry.h#L302-L319) function 

The best primer we have available in the source code for illustrating how the Settings Registry is used is the primary [ComponentApplication::MergeSettingsRegistry](https://github.com/o3de/o3de/blob/stabilization/2106/Code/Framework/AzCore/AzCore/Component/ComponentApplication.cpp#L927-L976) function.

That illustrates how to set specialization tags, as well as how to merge specific directories.

I am thinking that the BehaviorTree.Core gem could have in it's SystemComponent Activate function the logic to query the ModuleManager for the list of loaded gems via the `ModuleManager::EnumerateModules` function.

For each enumerated module it add a tag to specialization structure to only load that gem's settings registry file and then invokes the `SettingsRegistryInterface::MergeSettingsFolder` function

Here is psuedocode snippet of potentially doing this
```
BehaviorTreeSystemComponent::Activate()
{
    auto settingsRegistry = AZ::SettingsRegistry::Get();
    AZStd::vector<AzFramework::GemInfo> gemInfoList;
    if (AzFramework::GetGemsInfo(gemInfoList, *settingsRegistry))
    {
        // Get path to the BehaviorTree gem root
        auto FindBehaviorTreeGem = [](const AzFramework::GemInfo& gemInfo)
        {
            return gemInfo.m_gemName.starts_with("BehaviorTree");
        };

        AZ::IO::FixedMaxPath behaviorTreeGemRootPath;
        if (auto behaviorTreeGemFoundIt = AZStd::find_if(gemInfoList.begin(), gemInfoList.end(), FindBehaviorTreeGem);
            behaviorTreeGemFoundIt != behaviorTreeGemFoundIt)
        {
            behaviorTreeGemRootPath = behaviorTreeGemFoundIt->m_absoluteSourcePaths.front();
        }

        if (!behaviorTreeGemRootPath.empty())
        {
            auto GemEnumeratedCallback = [&gemInfoList, &settingsRegistry, &behaviorTreeGemRootPath](const AZ::ModuleData& moduleData) -> bool
            {
                if (AZ::DynamicModuleHandle* moduleHandle = moduleData.GetDynamicModuleHandle(); moduleHandle != nullptr)
                {
                    auto FindGemName = [&moduleHandle](const AzFramework::GemInfo& gemInfo)
                    {
                        // Use the stem of the Module file name. i.e For "EMotionFX.dll" use "EMotionFX"
                        AZ::IO::FixedMaxPathString moduleFileStem = AZ::IO::PathView(moduleHandle->GetFilename()).Stem().FixedMaxPathString();
                        return gemInfo.m_gemName == moduleFileStem;
                    };
                    if (auto gemFoundIt = AZStd::find_if(gemInfoList.begin(), gemInfoList.end(), FindGemName);
                        gemFoundIt != gemInfoList.end())
                    {
                        AZ::SettingsRegistryInterface::Specializations specializationTags{ gemFoundIt->m_gemName };
                        settingsRegistry->MergeSettingsFolder(behaviorTreeGemRootPath.Native(), specializationTags, AZ_TRAIT_OS_PLATFORM_CODENAME);
                    }
                }
                return true;
            };
            AZ::ModuleManagerRequestBus::Broadcast(&AZ::ModuleManagerRequests::EnumerateModules, GemEnumeratedCallback);
        }
    }

    // At this point all the settings for the loaded gems that the BehaviorTree Gem wishes to add nodes to are in the Settings Registry
    // The logic for registering the nodes can now be done
}
```

Reply:
Agree with @lumberyard-employee-dm here. If you want to know dynamically if a gem is loaded or not, that code that he is pointing at looks like a better approach.
Making it a decision at runtime is IMO the best option here because each of those optional gems act as "plugin" behavior to your system.

Reply:
Thanks @lumberyard-employee-dm for that great explanation, let me play a bit with it and I'll come back for feedback 🙏🏾. I didn't know about that setreg feature, and yeah it's not present in the docs 🤕

Reply:
We have docs for it, but... they are still in the process of being published.
There is an issue you can watch for it though: https://github.com/o3de/o3de.org/issues/706